federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+// replicas which can also lead to fast thrashing.
+// The only better way is to either ensure that object creation time stamp is set
+// and can be used authoritatively; or have another field on the local object
+// which is mandatorily set on creation and can be used authoritatively.
+// Should we abuse annotations again for this, or this can be a proper requirement?
+func isPristine(hpa *autoscalingv1.HorizontalPodAutoscaler) bool {
+	if hpa.Status.LastScaleTime == nil &&
+		hpa.Status.DesiredReplicas == 0 {
+		return true
+	}
+	return false
+}
+
+// isScaleable tells if it already has been a reasonable amount of
+// time since this hpa scaled. Its used to avoid fast thrashing.
+func isScaleable(hpa *autoscalingv1.HorizontalPodAutoscaler) bool {
+	if hpa.Status.LastScaleTime == nil {
+		return false
+	}
+	t := hpa.Status.LastScaleTime.Add(scaleForbiddenWindow)
+	if t.After(time.Now()) {
+		return false
+	}
+	return true
+}
+
+func maxReplicasReducible(obj pkgruntime.Object) bool {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	if (hpa.Spec.MinReplicas != nil) &&
+		(((hpa.Spec.MaxReplicas - 1) - *hpa.Spec.MinReplicas) < 0) {
+		return false
+	}
+	if isPristine(hpa) {
+		return true
+	}
+	if !isScaleable(hpa) {
+		return false
+	}
+	if (hpa.Status.DesiredReplicas < hpa.Status.CurrentReplicas) ||
+		((hpa.Status.DesiredReplicas == hpa.Status.CurrentReplicas) &&
+			(hpa.Status.DesiredReplicas < hpa.Spec.MaxReplicas)) {
+		return true
+	}
+	return false
+}
+
+// minReplicasReducible checks if this cluster (hpa) can offer replicas which are
+// stuck here because of min limit.
+// Its noteworthy, that min and max are adjusted separately, but if the replicas
+// are not being used here, the max adjustment will lead it to become equal to min,
+// but will not be able to scale down further and offer max to some other cluster
+// which needs replicas.
+func minReplicasReducible(obj pkgruntime.Object) bool {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	if isPristine(hpa) && (hpa.Spec.MinReplicas != nil) &&
+		(*hpa.Spec.MinReplicas > 1) &&
+		(*hpa.Spec.MinReplicas <= hpa.Spec.MaxReplicas) {
+		return true
+	}
+	if !isScaleable(hpa) {
+		return false
+	}
+	if (hpa.Spec.MinReplicas != nil) &&
+		(*hpa.Spec.MinReplicas > 1) &&
+		(hpa.Status.DesiredReplicas == hpa.Status.CurrentReplicas) &&
+		(hpa.Status.CurrentReplicas == *hpa.Spec.MinReplicas) {
+		return true
+	}
+	return false
+}
+
+func maxReplicasNeeded(obj pkgruntime.Object) bool {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	if !isScaleable(hpa) {
+		return false
+	}
+
+	if (hpa.Status.CurrentReplicas == hpa.Status.DesiredReplicas) &&
+		(hpa.Status.CurrentReplicas == hpa.Spec.MaxReplicas) {
+		return true
+	}
+	return false
+}
+
+func minReplicasAdjusteable(obj pkgruntime.Object) bool {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	if !isScaleable(hpa) ||
+		((hpa.Spec.MinReplicas != nil) &&
+			(*hpa.Spec.MinReplicas+1) > hpa.Spec.MaxReplicas) {
```

Comment:
Rather use >= and remove the '+1'?  
e.g. (*hpa.Spec.MinReplicas) >= hpa.Spec.MaxReplicas)